src/server/lib/okta/oidc.ts

```diff
@@ -0,0 +1,272 @@
+/* eslint-disable functional/immutable-data */
+import { Issuer, custom, IssuerMetadata } from 'openid-client';
+import { randomBytes } from 'crypto';
+import { Request } from 'express';
+import { getConfiguration } from '@/server/lib/getConfiguration';
+import { getProfileUrl } from '@/server/lib/getProfileUrl';
+import { Routes } from '@/shared/model/Routes';
+import { httpsAgent } from '@/server/lib/devHttpsAgent';
+import { ResponseWithRequestState } from '@/server/models/Express';
+import { logger } from '@/server/lib/logger';
+
+/**
+ * An object to hold the state the user was in while performing the
+ * Authorization Code flow.
+ * - `nonce` - Used as the `state` parameter in the Authorization Code Flow
+ *   to prevent CSRF attacks
+ * - `returnUrl` - The URL of the page/article/etc. that the user was on
+ *   before landing on the sign in page, so we can redirect them back there
+ *   once successfully authenticated
+ * @interface AuthorizationState
+ */
+interface AuthorizationState {
+  nonce: string;
+  returnUrl: string;
+}
+
+const { oktaDomain, oktaClientId, oktaClientSecret, oktaCustomOAuthServer } =
+  getConfiguration();
+
+// An agent to use with http calls to disable ssl verification in DEV only
+// for *.thegulocal.com domains
+custom.setHttpOptionsDefaults({
+  agent: {
+    https: httpsAgent,
+  },
+});
+
+// https://developer.okta.com/docs/reference/api/oidc/#well-known-openid-configuration
+// this is the metadata returned from the /.well-known/openid-configuration
+// used by openid-client to create the "Issuer" and "Client"
+const OIDC_METADATA: IssuerMetadata = {
```

Comment:
it would be good to create an oauth2 client which only supports the authorisation mechanisms and scopes we are using.